docs/architecture/adr-068-reverse-sync.md

```diff
@@ -0,0 +1,97 @@
+
+```go
+// Backfill fetches verifies and stores necessary history
+// to particapate in consensus / validate evidence
+func (r *Reactor) backfill(state State) error {}
+```
+
+`State` is used to work out how far to go back, namely we need all light blocks that have:
+- a height: `h >= state.LastBlockHeight - state.ConsensusParams.Evidence.MaxAgeNumBlocks`
+- a time: `t >= state.LastBlockTime - state.ConsensusParams.Evidence.MaxAgeDuration`
+
+Reverse Sync relies on two components: A `Dispatcher` and a `BlockQueue`. The `Dispatcher` is a pattern taken from a similar [PR](https://github.com/tendermint/tendermint/pull/4508). It is wired to the `LightBlockChannel` and allows for concurrent light block requests by shifting through a linked list of peers. This abstraction has the nice quality that it can also be used as an array of light providers for a P2P based light client.
+
+The `BlockQueue` is a data structure that allows for multiple workers to fetch light blocks, serializing them for the main thread which picks them off the end of the queue, verifies the hashes and persists them.
+
+### Integration with State Sync
+
+Reverse sync is a blocking process that runs directly after syncing state and before transitioning into either fast sync or consensus.
+
+Prior, the state sync service was not connected to any db, instead it passed the state and commit back to the node. For reverse sync, state sync will be given access to both the `StateStore` and `BlockStore` to be able to write `Header`'s, `Commit`'s and `ValidatorSet`'s and read them so as to serve other state syncing peers.
+
+This also means adding new methods to these respective stores in order to persist them
+
+### P2P Light Client Provider
+
+As mentioned previously, the `Dispatcher` is capable of handling requests to multiple peers. We can therefore simply peel off a `blockProvider` instance which is assigned to each peer. By giving it the chain ID, the `blockProvider` is capable of doing a basic validation of the light block before returning it to the client.
+
+It's important to note that because state sync doesn't have access to the evidence channel it is incapable of allowing the light client to report evidence thus `ReportEvidence` is a no op. This is not too much of a concern for reverse sync but will need to be addressed for pure p2p light clients.
+
+### Pruning
+
+A final small note is with pruning. This ADR will introduce changes that will not allow an application to prune blocks that are within the evidence age.
+
+## Future Work
+
+This ADR tries to remain within the scope of extending state sync, however the changes made opens the door for several areas to be followed up:
+- Properly integrate p2p messaging in the light client package. This will require adding the evidence channel so the light client is capable of reporting evidence. We may also need to rethink the providers model (i.e. currently providers are only added on start up)
```

Comment:
adding p2p to the light client? no everything is already there. It just needs to be wired up correctly

Reply:
tbh, I don't really follow this point :-/

Reply:
what is it exactly that's not so clear?

Reply:
I just don't understand it. What does integrate mean and what exactly are we integrating and why?

Reply:
At the moment light clients use RPC messages to get light blocks but now we are allowing light blocks to be passed via the p2p layer (because of state sync) so by "integrate", I mean to allow light clients to use the P2P stack to get the light blocks